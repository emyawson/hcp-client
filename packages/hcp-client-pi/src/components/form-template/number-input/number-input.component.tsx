import { InputContent } from '@roche/patterns-indicators/components/form-template/form-template.style';
import * as React from 'react';

import {
  NumberValidation,
  PatternValidation,
} from '../../../types/config.types';
import { ChangeHandler, NumberInputProps } from '../form-template.types';

class NumberInput extends React.Component<
  NumberInputProps,
  { error: boolean }
> {
  constructor(props: NumberInputProps) {
    super(props);
    this.state = { error: false };
  }

  public render() {
    const { name, path, value } = this.props;
    return (
      <InputContent
        error={this.state.error}
        type="number"
        name={name}
        key={`number-input-key-${path}-${name}`}
        defaultValue={`${value}`}
        onChange={this.makeChangeHandler(path, name)}
      />
    );
  }

  public makeChangeHandler(path, fieldName) {
    return (event: React.ChangeEvent<any>) => {
      const numberInputValue = event.target.value;
      if (this.props.validationConfig[fieldName].type === 'number') {
        const validationConfig = this.props.validationConfig[
          fieldName
        ] as NumberValidation & { required: boolean };

        if (
          numberInputValue < validationConfig.min ||
          numberInputValue > validationConfig.max
        ) {
          this.setState({ error: true });
        } else {
          this.setState({ error: false });
        }
      }

      // tslint:disable-next-line:no-unused-expression
      this.props.changeHandler &&
        this.props.changeHandler(path, fieldName, Number(event.target.value));
    };
  }
}

export const NumberInputFactory = (
  onChange: ChangeHandler,
  validationConfig: PatternValidation,
) => (N: NumberInputProps) => (
  <NumberInput
    key={`number-input-${N.path}-${N.name}`}
    {...N}
    changeHandler={onChange}
    validationConfig={validationConfig}
  />
);
