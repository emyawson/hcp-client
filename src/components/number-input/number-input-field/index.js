export { NumberInputField } from './number-input-field.component';
