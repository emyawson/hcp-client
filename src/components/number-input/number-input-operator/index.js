export * from './number-input-operator.component';
