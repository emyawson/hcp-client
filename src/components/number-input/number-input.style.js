import styled from 'styled-components';

import { borderRadius, colors, fontSize } from 'src/core';

import { NumberInputFieldControl } from './number-input-field/number-input-field.style';

const colorBorder = colors.grayLight;

export const NumberInputContainer = styled.div`
  background: ${colors.white};
  border-radius: ${borderRadius.three};
  border: 1px solid ${colorBorder};
  font-size: ${fontSize.p};
  display: inline-flex;
  overflow: hidden;

  ${NumberInputFieldControl} {
    border: 1px solid transparent;
    border-radius: 0;
  }
`;

export const NumberInputOperatorWrapper = styled.div`
  border-left: 1px solid ${colorBorder};
  display: flex;
  flex-direction: column;
`;

export const OperatorDivider = styled.div`
  background-color: ${colorBorder};
  height: 1px;
  width: 100%;
`;

// Vertically center subtract icon with the PlusIcon
export const MinusIconSpan = styled.span`
  display: inline-block;
  transform: translateY(-3px);
`;
