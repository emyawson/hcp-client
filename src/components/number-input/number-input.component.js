import React, { Component } from 'react';
import { isEmpty } from 'ramda';

import { RenderIf } from 'src/utils';
import { PlusIcon, MinusIcon } from 'src/assets/icons';

import { NumberInputField } from './number-input-field';
import {
  NumberInputContainer,
  NumberInputOperatorWrapper,
  OperatorDivider,
  MinusIconSpan,
} from './number-input.style';
import { NumberInputOperator } from './number-input-operator';

const addButtonIcon = <PlusIcon height={10} />;
const subtractButtonIcon = (
  <MinusIconSpan>
    <MinusIcon />
  </MinusIconSpan>
);

export class NumberInput extends Component {
  constructor(props) {
    super(props);
    // The current model value of the number is stored
    // in component state so that the increment and
    // decrement functions can access the value
    const { initialValue } = this.props;
    this.state = {
      modelValue: initialValue,
    };
  }

  render() {
    const { id, model, min, max, disabled = false, row = false } = this.props;
    const numberInputField = (
      <NumberInputField
        id={id}
        model={model}
        min={min}
        max={max}
        disabled={disabled}
        updateField={this.updateField}
      />
    );
    return (
      <NumberInputContainer disabled={disabled}>
        <RenderIf validate={row}>
          <NumberInputOperator
            icon={subtractButtonIcon}
            operatorAction={this.decrement}
            position="left"
            disabled={disabled}
          />
          {numberInputField}
          <NumberInputOperator
            icon={addButtonIcon}
            operatorAction={this.increment}
            position="right"
            disabled={disabled}
          />
        </RenderIf>
        <RenderIf validate={!row}>
          {numberInputField}
          <NumberInputOperatorWrapper>
            <NumberInputOperator
              icon={addButtonIcon}
              operatorAction={this.increment}
              disabled={disabled}
            />
            <OperatorDivider />
            <NumberInputOperator
              icon={subtractButtonIcon}
              operatorAction={this.decrement}
              disabled={disabled}
            />
          </NumberInputOperatorWrapper>
        </RenderIf>
      </NumberInputContainer>
    );
  }

  // updateField: updates the RRF local form model using updateValue
  // as well as this component's state (needed for the increment/decrement).
  // Bundled together so they are always in sync.
  //
  // updateValue should be a class method belonging to the parent
  // component (where the local form is located)
  // It should call RRF's actions.change to update the local form state
  updateField = value => {
    const { updateValue } = this.props;
    const parsedValue = isEmpty(value) ? '' : parseInt(value, 10);
    this.setState({ modelValue: parsedValue });
    updateValue(parsedValue);
  };

  increment = () => {
    const { max } = this.props;
    const { modelValue } = this.state;
    const newValue = modelValue >= max ? max : modelValue + 1;
    this.updateField(newValue);
  };

  decrement = () => {
    const { min } = this.props;
    const { modelValue } = this.state;
    const newValue = modelValue <= min ? min : modelValue - 1;
    this.updateField(newValue);
  };
}

NumberInput.defaultProps = {
  max: 99,
  min: 0,
};
