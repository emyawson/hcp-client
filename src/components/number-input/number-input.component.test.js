import React from 'react';
import { shallow } from 'enzyme';

import { NumberInput } from './number-input.component';

describe('Number input component', () => {
  const baseProps = {
    initialValue: 0,
    id: 'testInput',
    model: 'numberOfStrips',
    min: 0,
    max: 99,
  };

  it('renders correctly', () => {
    const wrapper = shallow(<NumberInput {...baseProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls the given value update function and updates state with parsed int value', () => {
    const mockFn = jest.fn();
    const mockProps = {
      ...baseProps,
      updateValue: mockFn,
    };
    const wrapper = shallow(<NumberInput {...mockProps} />);
    wrapper.instance().updateField('88');
    expect(mockFn).toBeCalledWith(88);
    expect(wrapper.state('modelValue')).toEqual(88);
  });

  it('updates model value with empty string', () => {
    const mockProps = {
      ...baseProps,
      updateValue: () => null,
    };
    const wrapper = shallow(<NumberInput {...mockProps} />);
    wrapper.instance().updateField('');
    expect(wrapper.state('modelValue')).toEqual('');
  });

  it('increments model value in state', () => {
    const mockProps = {
      ...baseProps,
      initialValue: 2,
      updateValue: () => {},
    };
    const wrapper = shallow(<NumberInput {...mockProps} />);
    wrapper.instance().increment();
    expect(wrapper.state('modelValue')).toEqual(3);
  });

  it('sets value to max if model value is equal or greater to max', () => {
    const mockProps = {
      ...baseProps,
      initialValue: 10,
      max: 10,
      updateValue: () => {},
    };
    const wrapper = shallow(<NumberInput {...mockProps} />);
    wrapper.instance().increment();
    expect(wrapper.state('modelValue')).toEqual(10);
  });

  it('decrements model value in state', () => {
    const mockProps = {
      ...baseProps,
      initialValue: 2,
      updateValue: () => {},
    };
    const wrapper = shallow(<NumberInput {...mockProps} />);
    wrapper.instance().decrement();
    expect(wrapper.state('modelValue')).toEqual(1);
  });

  it('sets value to min if model value is equal or less than min', () => {
    const mockProps = {
      ...baseProps,
      initialValue: 2,
      min: 2,
      updateValue: () => {},
    };
    const wrapper = shallow(<NumberInput {...mockProps} />);
    wrapper.instance().decrement();
    expect(wrapper.state('modelValue')).toEqual(2);
  });
});
