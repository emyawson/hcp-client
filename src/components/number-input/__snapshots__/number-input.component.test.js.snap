// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number input component renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NumberInput
    id="testInput"
    initialValue={0}
    max={99}
    min={0}
    model="numberOfStrips"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <f1
          validate={false}
        >
          <Unknown
            disabled={false}
            icon={
              <styled.span>
                <Unknown />
              </styled.span>
            }
            operatorAction={[Function]}
            position="left"
          />
          <NumberInputField
            disabled={false}
            id="testInput"
            max={99}
            min={0}
            model="numberOfStrips"
            updateField={[Function]}
          />
          <Unknown
            disabled={false}
            icon={
              <Unknown
                height={10}
              />
            }
            operatorAction={[Function]}
            position="right"
          />
        </f1>,
        <f1
          validate={true}
        >
          <NumberInputField
            disabled={false}
            id="testInput"
            max={99}
            min={0}
            model="numberOfStrips"
            updateField={[Function]}
          />
          <styled.div>
            <Unknown
              disabled={false}
              icon={
                <Unknown
                  height={10}
                />
              }
              operatorAction={[Function]}
            />
            <styled.div />
            <Unknown
              disabled={false}
              icon={
                <styled.span>
                  <Unknown />
                </styled.span>
              }
              operatorAction={[Function]}
            />
          </styled.div>
        </f1>,
      ],
      "disabled": false,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Unknown
              disabled={false}
              icon={
                <styled.span>
                  <Unknown />
                </styled.span>
              }
              operatorAction={[Function]}
              position="left"
            />,
            <NumberInputField
              disabled={false}
              id="testInput"
              max={99}
              min={0}
              model="numberOfStrips"
              updateField={[Function]}
            />,
            <Unknown
              disabled={false}
              icon={
                <Unknown
                  height={10}
                />
              }
              operatorAction={[Function]}
              position="right"
            />,
          ],
          "validate": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "function",
            "props": Object {
              "disabled": false,
              "icon": <styled.span>
                <Unknown />
              </styled.span>,
              "operatorAction": [Function],
              "position": "left",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "function",
            "props": Object {
              "disabled": false,
              "id": "testInput",
              "max": 99,
              "min": 0,
              "model": "numberOfStrips",
              "updateField": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "function",
            "props": Object {
              "disabled": false,
              "icon": <Unknown
                height={10}
              />,
              "operatorAction": [Function],
              "position": "right",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <NumberInputField
              disabled={false}
              id="testInput"
              max={99}
              min={0}
              model="numberOfStrips"
              updateField={[Function]}
            />,
            <styled.div>
              <Unknown
                disabled={false}
                icon={
                  <Unknown
                    height={10}
                  />
                }
                operatorAction={[Function]}
              />
              <styled.div />
              <Unknown
                disabled={false}
                icon={
                  <styled.span>
                    <Unknown />
                  </styled.span>
                }
                operatorAction={[Function]}
              />
            </styled.div>,
          ],
          "validate": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "function",
            "props": Object {
              "disabled": false,
              "id": "testInput",
              "max": 99,
              "min": 0,
              "model": "numberOfStrips",
              "updateField": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Unknown
                  disabled={false}
                  icon={
                    <Unknown
                      height={10}
                    />
                  }
                  operatorAction={[Function]}
                />,
                <styled.div />,
                <Unknown
                  disabled={false}
                  icon={
                    <styled.span>
                      <Unknown />
                    </styled.span>
                  }
                  operatorAction={[Function]}
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": null,
                "nodeType": "function",
                "props": Object {
                  "disabled": false,
                  "icon": <Unknown
                    height={10}
                  />,
                  "operatorAction": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "function",
                "props": Object {
                  "disabled": false,
                  "icon": <styled.span>
                    <Unknown />
                  </styled.span>,
                  "operatorAction": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <f1
            validate={false}
          >
            <Unknown
              disabled={false}
              icon={
                <styled.span>
                  <Unknown />
                </styled.span>
              }
              operatorAction={[Function]}
              position="left"
            />
            <NumberInputField
              disabled={false}
              id="testInput"
              max={99}
              min={0}
              model="numberOfStrips"
              updateField={[Function]}
            />
            <Unknown
              disabled={false}
              icon={
                <Unknown
                  height={10}
                />
              }
              operatorAction={[Function]}
              position="right"
            />
          </f1>,
          <f1
            validate={true}
          >
            <NumberInputField
              disabled={false}
              id="testInput"
              max={99}
              min={0}
              model="numberOfStrips"
              updateField={[Function]}
            />
            <styled.div>
              <Unknown
                disabled={false}
                icon={
                  <Unknown
                    height={10}
                  />
                }
                operatorAction={[Function]}
              />
              <styled.div />
              <Unknown
                disabled={false}
                icon={
                  <styled.span>
                    <Unknown />
                  </styled.span>
                }
                operatorAction={[Function]}
              />
            </styled.div>
          </f1>,
        ],
        "disabled": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Unknown
                disabled={false}
                icon={
                  <styled.span>
                    <Unknown />
                  </styled.span>
                }
                operatorAction={[Function]}
                position="left"
              />,
              <NumberInputField
                disabled={false}
                id="testInput"
                max={99}
                min={0}
                model="numberOfStrips"
                updateField={[Function]}
              />,
              <Unknown
                disabled={false}
                icon={
                  <Unknown
                    height={10}
                  />
                }
                operatorAction={[Function]}
                position="right"
              />,
            ],
            "validate": false,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "disabled": false,
                "icon": <styled.span>
                  <Unknown />
                </styled.span>,
                "operatorAction": [Function],
                "position": "left",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "disabled": false,
                "id": "testInput",
                "max": 99,
                "min": 0,
                "model": "numberOfStrips",
                "updateField": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "disabled": false,
                "icon": <Unknown
                  height={10}
                />,
                "operatorAction": [Function],
                "position": "right",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <NumberInputField
                disabled={false}
                id="testInput"
                max={99}
                min={0}
                model="numberOfStrips"
                updateField={[Function]}
              />,
              <styled.div>
                <Unknown
                  disabled={false}
                  icon={
                    <Unknown
                      height={10}
                    />
                  }
                  operatorAction={[Function]}
                />
                <styled.div />
                <Unknown
                  disabled={false}
                  icon={
                    <styled.span>
                      <Unknown />
                    </styled.span>
                  }
                  operatorAction={[Function]}
                />
              </styled.div>,
            ],
            "validate": true,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "function",
              "props": Object {
                "disabled": false,
                "id": "testInput",
                "max": 99,
                "min": 0,
                "model": "numberOfStrips",
                "updateField": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <Unknown
                    disabled={false}
                    icon={
                      <Unknown
                        height={10}
                      />
                    }
                    operatorAction={[Function]}
                  />,
                  <styled.div />,
                  <Unknown
                    disabled={false}
                    icon={
                      <styled.span>
                        <Unknown />
                      </styled.span>
                    }
                    operatorAction={[Function]}
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "function",
                  "props": Object {
                    "disabled": false,
                    "icon": <Unknown
                      height={10}
                    />,
                    "operatorAction": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "class",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "function",
                  "props": Object {
                    "disabled": false,
                    "icon": <styled.span>
                      <Unknown />
                    </styled.span>,
                    "operatorAction": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
